### PR TITLE
feat: Shareable resource

### PR DESCRIPTION
**Pull request by josebui**
## Description

- Shareable resource file for Landscapes and groups

### Checklist
- [x] Corresponding issue has been opened
- [x] New tests added

### Related Issues
- https://github.com/techmatters/terraso-web-client/issues/1322

### CODE REVIEW

File: terraso_backend/apps/graphql/schema/shared_resources.py [hunk 1]
```diff
@@ -51,3 +62,85 @@ def resolve_source(self, info, **kwargs):
 
     def resolve_target(self, info, **kwargs):
         return self.target
+
+    def resolve_download_url(self, info, **kwargs):
+        return f"{settings.API_ENDPOINT}/shared-data/download/{self.share_uuid}"
+
+    def resolve_share_url(self, info, **kwargs):
+        target = self.target
+        entity = "groups" if isinstance(target, Group) else "landscapes"
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/shared_resources.py:71]: issue: do an `isinstance` check for `Landscape` too.

File: terraso_backend/tests/conftest.py [hunk 2]
```diff
@@ -46,6 +46,11 @@ def logged_client(access_token):
     return Client(HTTP_AUTHORIZATION=f"Bearer {access_token}")
 
 
+@pytest.fixture
+def not_logged_client(access_token):
```
Comment (paulschreiber) [inline terraso_backend/tests/conftest.py:50]: question: what does `not_logged` mean?

File: terraso_backend/apps/core/models/shared_resources.py [hunk 3]
```diff
@@ -26,6 +29,17 @@ class SharedResource(BaseModel):
     Target represents the resource that is receiving the shared resource (Example: Landscape).
     """
 
+    SHARE_ACCESS_NO = "no"
```
Comment (paulschreiber) [inline terraso_backend/apps/core/models/shared_resources.py:32]: suggestion: call this `SHARE_ACCESS_NONE`.

File: terraso_backend/tests/conftest.py (hunk 2 above)
Comment (josebui) [reply terraso_backend/tests/conftest.py:50]: @paulschreiber mean that it is a fixture for a client to generate requests simulating a request from a not logged in user

Comment (paulschreiber) [reply terraso_backend/tests/conftest.py:50]: Ah. Please call it `not_logged_in_client` then.

File: terraso_backend/apps/core/models/shared_resources.py [hunk 4]
```diff
@@ -26,6 +29,15 @@ class SharedResource(BaseModel):
     Target represents the resource that is receiving the shared resource (Example: Landscape).
     """
 
+    SHARE_ACCESS_ALL = "all"
+    SHARE_ACCESS_TARGET_MEMBERS = "target_members"
+    DEFAULT_SHARE_ACCESS = SHARE_ACCESS_TARGET_MEMBERS
+
+    SHARE_ACCESS_TYPES = (
+        (SHARE_ACCESS_ALL, _("Anyone with the link")),
+        (SHARE_ACCESS_TARGET_MEMBERS, _("Only target members")),
```
Comment (paulschreiber) [inline terraso_backend/apps/core/models/shared_resources.py:38]:
```suggestion
    SHARE_ACCESS_ALL = "all"
    SHARE_ACCESS_MEMBERS = "target_members"
    DEFAULT_SHARE_ACCESS = SHARE_ACCESS_MEMBERS

    SHARE_ACCESS_TYPES = (
        (SHARE_ACCESS_ALL, _("Anyone with the link")),
        (SHARE_ACCESS_MEMBERS, _("Only members")),
```

File: terraso_backend/apps/core/models/shared_resources.py [hunk 5]
```diff
@@ -37,3 +49,27 @@ class SharedResource(BaseModel):
...
+                fields=("share_uuid",),
+                condition=models.Q(deleted_at__isnull=True),
+                name="unique_share_uuid",
+            ),
+        )
+
+    @classmethod
+    def get_share_access_from_text(cls, share_access):
+        if not share_access:
+            return cls.SHARE_ACCESS_TARGET_MEMBERS
```
Comment (paulschreiber) [inline terraso_backend/apps/core/models/shared_resources.py:71]:
```suggestion
            return cls.SHARE_ACCESS_MEMBERS
```

File: terraso_backend/apps/core/models/shared_resources.py [hunk 6]
```diff
@@ -37,3 +49,27 @@ class SharedResource(BaseModel):
...
+        lowered = share_access.lower()
+        if lowered == cls.SHARE_ACCESS_ALL:
+            return cls.SHARE_ACCESS_ALL
+        return cls.SHARE_ACCESS_TARGET_MEMBERS
```
Comment (paulschreiber) [inline terraso_backend/apps/core/models/shared_resources.py:75]:
```suggestion
        return cls.SHARE_ACCESS_MEMBERS
```

File: terraso_backend/apps/graphql/schema/shared_resources.py [hunk 7]
```diff
@@ -51,3 +62,90 @@ def resolve_source(self, info, **kwargs):
...
+        Landscape.objects.filter(
+            membership_list__memberships__deleted_at__isnull=True,
+            membership_list__memberships__user__id=user_pk,
+            membership_list__memberships__membership_status=CollaborationMembership.APPROVED,
+        ).values("id")
+    )
+
+    share_access_all = Q(share_access=SharedResource.SHARE_ACCESS_ALL)
+    share_access_members = Q(
+        Q(share_access=SharedResource.SHARE_ACCESS_TARGET_MEMBERS)
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/shared_resources.py:111]:
```suggestion
        Q(share_access=SharedResource.SHARE_ACCESS_MEMBERS)
```

File: terraso_backend/apps/shared_data/permission_rules.py [hunk 8]
```diff
@@ -103,4 +103,16 @@ def allowed_to_delete_visualization_config(user, visualization_config):
...
 
 
+@rules.predicate
+def allowed_to_download_data_entry_file(user, shared_resource):
+    target = shared_resource.target
+
+    if shared_resource.share_access == SharedResource.SHARE_ACCESS_ALL:
+        return True
+
+    if shared_resource.share_access == SharedResource.SHARE_ACCESS_TARGET_MEMBERS:
```
Comment (paulschreiber) [inline terraso_backend/apps/shared_data/permission_rules.py:113]:
```suggestion
    if shared_resource.share_access == SharedResource.SHARE_ACCESS_MEMBERS:
```

File: terraso_backend/tests/graphql/test_shared_resource.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    data_entry = data_entries[0]
+    shared_resource = data_entry.shared_resources.all()[0]
+
+    shared_resource.target.membership_list.memberships.all().delete()
+
+    shared_resource.target.membership_list.save_membership(
+        users[0].email, group_collaboration_roles.ROLE_MEMBER, CollaborationMembership.APPROVED
+    )
+
+    shared_resource.share_access = SharedResource.SHARE_ACCESS_TARGET_MEMBERS
```
Comment (paulschreiber) [inline terraso_backend/tests/graphql/test_shared_resource.py:61]:
```suggestion
    shared_resource.share_access = SharedResource.SHARE_ACCESS_MEMBERS
```

File: terraso_backend/tests/graphql/test_shared_resource.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    assert shared_resource.share_access == result["shareAccess"].lower()
+
+
+def test_shared_resource_access_members_fail(client_query, data_entries, users):
+    data_entry = data_entries[0]
+    shared_resource = data_entry.shared_resources.all()[0]
+
+    shared_resource.target.membership_list.memberships.all().delete()
+
+    shared_resource.share_access = SharedResource.SHARE_ACCESS_TARGET_MEMBERS
```
Comment (paulschreiber) [inline terraso_backend/tests/graphql/test_shared_resource.py:86]:
```suggestion
    shared_resource.share_access = SharedResource.SHARE_ACCESS_MEMBERS
```

File: terraso_backend/tests/shared_data/conftest.py [hunk 11]
```diff
@@ -117,3 +121,53 @@ def visualization_config_gpx(user):
...
+@pytest.fixture
+def shared_resource_data_entry_shared_target_members(users):
+    creator = users[0]
+    creator_group = mixer.blend(Group)
+    creator_group.membership_list.save_membership(
+        creator.email, group_collaboration_roles.ROLE_MEMBER, CollaborationMembership.APPROVED
+    )
+    return mixer.blend(
+        SharedResource,
+        share_access=SharedResource.SHARE_ACCESS_TARGET_MEMBERS,
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/conftest.py:149]:
```suggestion
        share_access=SharedResource.SHARE_ACCESS_MEMBERS,
```

File: terraso_backend/tests/shared_data/conftest.py [hunk 12]
```diff
@@ -117,3 +121,53 @@ def visualization_config_gpx(user):
...
+        share_uuid=uuid.uuid4(),
+        target=mixer.blend(Group),
+        source=mixer.blend(
+            DataEntry, slug=None, created_by=creator, size=100, entry_type=DataEntry.ENTRY_TYPE_FILE
+        ),
+    )
+
+
+@pytest.fixture
+def shared_resource_data_entry_shared_target_members(users):
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/conftest.py:141]:
```suggestion
def shared_resource_data_entry_shared_members(users):
```

File: terraso_backend/tests/shared_data/conftest.py [hunk 13]
```diff
@@ -117,3 +121,53 @@ def visualization_config_gpx(user):
...
+        share_uuid=uuid.uuid4(),
+        target=creator_group,
+        source=mixer.blend(
+            DataEntry, slug=None, created_by=creator, size=100, entry_type=DataEntry.ENTRY_TYPE_FILE
+        ),
+    )
+
+
+@pytest.fixture
+def shared_resource_data_entry_shared_target_members_user_1(users):
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/conftest.py:159]:
```suggestion
def shared_resource_data_entry_shared_members_user_1(users):
```

File: terraso_backend/tests/shared_data/conftest.py [hunk 14]
```diff
@@ -117,3 +121,53 @@ def visualization_config_gpx(user):
...
+    creator = users[1]
+    creator_group = mixer.blend(Group)
+    creator_group.membership_list.save_membership(
+        creator.email, group_collaboration_roles.ROLE_MEMBER, CollaborationMembership.APPROVED
+    )
+    return mixer.blend(
+        SharedResource,
+        share_access=SharedResource.SHARE_ACCESS_TARGET_MEMBERS,
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/conftest.py:167]:
```suggestion
        share_access=SharedResource.SHARE_ACCESS_MEMBERS,
```

File: terraso_backend/tests/shared_data/test_views.py [hunk 15]
```diff
@@ -165,3 +165,53 @@ def test_create_data_entry_file_invalid_type(logged_client, upload_url, data_ent
...
+    )
+    response = not_logged_in_client.get(url)
+
+    assert response.status_code == 302
+    assert response.url == redirect_url
+
+
+@mock.patch("apps.shared_data.models.data_entries.data_entry_file_storage.url")
+def test_download_data_entry_file_shared_target_members(
+    get_url_mock, logged_client, shared_resource_data_entry_shared_target_members
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/test_views.py:188]:
```suggestion
def test_download_data_entry_file_shared_members(
    get_url_mock, logged_client, shared_resource_data_entry_shared_members
```

File: terraso_backend/tests/shared_data/test_views.py [hunk 16]
```diff
@@ -165,3 +165,53 @@ def test_create_data_entry_file_invalid_type(logged_client, upload_url, data_ent
...
+):
+    redirect_url = "https://example.org/s3_file.json"
+    get_url_mock.return_value = redirect_url
+    url = reverse(
+        "shared_data:download",
+        kwargs={
+            "shared_resource_uuid": shared_resource_data_entry_shared_target_members.share_uuid
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/test_views.py:195]:
```suggestion
            "shared_resource_uuid": shared_resource_data_entry_shared_members.share_uuid
```

File: terraso_backend/tests/shared_data/test_views.py [hunk 17]
```diff
@@ -165,3 +165,53 @@ def test_create_data_entry_file_invalid_type(logged_client, upload_url, data_ent
...
+    )
+    response = logged_client.get(url)
+
+    assert response.status_code == 302
+    assert response.url == redirect_url
+
+
+@mock.patch("apps.shared_data.models.data_entries.data_entry_file_storage.url")
+def test_download_data_entry_file_shared_target_members_fail(
+    get_url_mock, logged_client, shared_resource_data_entry_shared_target_members_user_1
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/test_views.py:206]:
```suggestion
    get_url_mock, logged_client, shared_resource_data_entry_shared_members_user_1
```

File: terraso_backend/tests/shared_data/test_views.py [hunk 18]
```diff
@@ -165,3 +165,53 @@ def test_create_data_entry_file_invalid_type(logged_client, upload_url, data_ent
...
+):
+    redirect_url = "https://example.org/s3_file.json"
+    get_url_mock.return_value = redirect_url
+    share_uuid = shared_resource_data_entry_shared_target_members_user_1.share_uuid
```
Comment (paulschreiber) [inline terraso_backend/tests/shared_data/test_views.py:210]:
```suggestion
    share_uuid = shared_resource_data_entry_shared_members_user_1.share_uuid
```

File: terraso_backend/apps/graphql/schema/shared_resources.py [hunk 19]
```diff
@@ -51,3 +62,90 @@ def resolve_source(self, info, **kwargs):
 
     def resolve_target(self, info, **kwargs):
         return self.target
+
+    def resolve_download_url(self, info, **kwargs):
+        return f"{settings.API_ENDPOINT}/shared-data/download/{self.share_uuid}"
```
Comment (paulschreiber) [inline terraso_backend/apps/graphql/schema/shared_resources.py:67]:
```suggestion
        return f"{settings.API_ENDPOINT}/download/{self.share_uuid}"
```

File: terraso_backend/tests/graphql/test_shared_data.py [hunk 20]
```diff
@@ -417,10 +414,28 @@ def test_data_entries_from_parent_query_by_resource_field(
...
+            "download_url": resource["node"]["downloadUrl"],
+            "share_access": resource["node"]["shareAccess"],
+        }
+        for resource in resources
+    ]
 
     for data_entry in data_entries:
-        assert data_entry.name in entries_result
+        uuid = data_entry.shared_resources.all()[0].share_uuid
+        download_url = f"{settings.API_ENDPOINT}/shared-data/download/{uuid}"
```
Comment (paulschreiber) [inline terraso_backend/tests/graphql/test_shared_data.py:429]: issue: a helper like `resolve_download_url` every time you generate URLs of this form.

Comment (paulschreiber) [inline terraso_backend/tests/graphql/test_shared_data.py:429]:
```suggestion
        download_url = f"{settings.API_ENDPOINT}/download/{uuid}"
```

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: terraso_backend/apps/graphql/schema/shared_resources.py (hunk 19 above)
Comment (josebui) [reply terraso_backend/apps/graphql/schema/shared_resources.py:67]: @paulschreiber This is the API endpoint, the urls are organized by app, so we can't avoid the `shared-data` without losing that structure

File: terraso_backend/tests/graphql/test_shared_resource.py [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+    data_entry = data_entries[0]
+    shared_resource = data_entry.shared_resources.all()[0]
+
+    shared_resource.target.membership_list.memberships.all().delete()
+
+    shared_resource.share_access = SharedResource.SHARE_ACCESS_ALL
+    shared_resource.save()
+
+    response = client_query(
+        """
```
Comment (paulschreiber) [inline terraso_backend/tests/graphql/test_shared_resource.py:36]: question: would an f-string make sense here?

Comment (paulschreiber) [reply terraso_backend/tests/graphql/test_shared_resource.py:36]: Okay — let's make a task to do that. We can then add a flak8 plugin to enforce it. https://github.com/MichaelKim0407/flake8-use-fstring

**Pull request merged**